File: 2015/day20/day20.js
const fs = require("fs");

class House {
  getInput() {
    const inputFile = this.useTest ? "input-test.txt" : "input.txt";
    try {
      return parseInt(fs.readFileSync(inputFile, "utf8").trim().split("\r\n")[0]);
    } catch (err) {
      throw err;
    }
  }

  constructor(useTest = false) {
    this.useTest = useTest;
    this.target = this.getInput();
  }

  getlowestNumber(isPart2) {
    const multipler = isPart2 ? 11 : 10;
    let curr = 1;
    let currSum = 0;
    while (currSum < this.target ) {
      currSum = 0;
      for (let i = 1; i <= Math.ceil(Math.sqrt(curr)); i++) {
        if (curr % i === 0) {
          if (isPart2) {
            if (i * 50 > curr) currSum += multipler * i;
            if (i !== curr / i && (curr / i) * 50 > curr) currSum += (multipler * curr) / i;
          } else {
            currSum += multipler * i;
            if (i !== curr / i) currSum += (multipler * curr) / i;
          }
        }
      }
      curr += 1;
    }
    return curr - 1;
  }
}

const house = new House();
console.log("Day 20 part 1:", house.getlowestNumber());
console.log("Day 20 part 2:", house.getlowestNumber(true));
// Total Runtime ~1.9s

File: 2023/day20/day20.py
from typing import Optional, Tuple
from collections import deque, defaultdict
from math import lcm


class Pulse:
    def getInput(self) -> Tuple:
        inputFile = "input-test.txt" if self.useTest else "input.txt"
        broadcast = []
        flip, conj = {}, {}
        with open(inputFile, "r") as file1:
            for line in file1:
                line = line.strip()
                if line.startswith("broadcaster"):
                    broadcast = line.split(" -> ")[1].split(", ")
                    continue
                if line.startswith("%"):
                    split = line.split(" -> ")
                    flip[split[0][1:]] = split[1].split(", ")
                else:
                    split = line.split(" -> ")
                    conj[split[0][1:]] = split[1].split(", ")
            return broadcast, flip, conj

    def __init__(self, useTest: Optional[bool] = False) -> None:
        self.useTest = useTest
        self.broadast, self.flip, self.conj = self.getInput()

    def reset(self) -> None:
        self.val = {}
        self.Cin = defaultdict(list)
        for key, outputs in [*self.conj.items(), *self.flip.items()]:
            self.val[key] = 0
            for n in outputs:
                if n in self.conj:
                    self.Cin[n].append(key)

    def broadcast(
        self,
        feed: Optional[str] = "",
        presses: Optional[int] = 0,
    ) -> None:
        signals = [(x, 0) for x in self.broadast]
        q = deque(signals)
        self.counts[0] += 1
        output = "output" if self.useTest else "rx"
        while q:
            curr, signal = q.popleft()
            self.counts[signal] += 1
            if curr == output:
                continue
            if curr in self.flip:
                if signal:
                    continue
                self.val[curr] = 1 if not self.val[curr] else 0
                s = [(x, self.val[curr]) for x in self.flip[curr]]
            elif curr in self.conj:
                self.val[curr] = 0 if all(self.val[x] for x in self.Cin[curr]) else 1
                if feed and feed in self.conj[curr] and self.val[curr]:
                    self.cycles[curr] = presses
                s = [(x, self.val[curr]) for x in self.conj[curr]]
            q.extend(s)

    def getTotal(self) -> int:
        self.reset()
        self.counts = [0, 0]
        for _ in range(1000):
            self.broadcast()
        return self.counts[0] * self.counts[1]

    def getButtonPress(self) -> int:
        self.reset()
        # Assumed feed is a conjunction
        for key, outputs in self.conj.items():
            if "rx" in outputs:
                feed = key
                break
        self.cycles = {name: None for name in self.Cin[feed]}
        counter = 0
        while True:
            counter += 1
            self.broadcast(feed, counter)
            if all(self.cycles[key] for key in self.Cin[feed]):
                break
        return lcm(*self.cycles.values())


if __name__ == "__main__":
    pulse = Pulse()
    print("Day 20 part 1:", pulse.getTotal())
    print("Day 20 part 2:", pulse.getButtonPress())
